src/fe/inf_fe_static.C

```diff
@@ -1374,10 +1383,25 @@ void InfFE<Dim, T_radial, T_map>::inf_compute_constraints (DofConstraints & cons
       if (self_constraint)
         continue;
 
+      // if the base is not coarser, than only those DOFS shared with respective neighbors need constraints:
+      if (coarser_neighbors[0] != 0)
+        {
+          bool needs_account=false;
+          for (auto & X : coarser_neighbors)
+            if ( child_elem->neighbor_ptr(X)->contains_point(child_side->point(child_side_dof)) )
```

Comment:
I'm having trouble following what the coarser_neighbors code is supposed to do (why does any elem-neighbor pair care about the other neighbors' status?), but this part in particular is greatly confusing me.  You're passing child_side_dof, a dof index, into a function expecting a point index?  What happens if you have more DoFs than points?

Reply:
I agree that this is not written well; I'll try to rework it to be more readable. 
> why does any elem-neighbor pair care about the other neighbors' status?

I think, there are two issues: First, I tried some shortcut in the logic of the function; but while trying to write this down, I see that there is too much confinement. So thanks for pointing me there!

> You're passing child_side_dof, a dof index, into a function expecting a point index?
Probably I missed something here: I adapted this from `fe_lagrange.C:803`:
````
              // The support point of the DOF
              const Point & support_point = my_side->point(my_dof);
````
which, to my mind, is the very same: `my_dof` is the dof index of `my_side` which is used to obtain its corresponding support point.
Is there another, safer, way to obtain the point that I want to get?

Reply:
I usually see the "one dof index" == "one point" confusion in people used to working with Lagrange elements, where that actually holds.  But libMesh has a ton of other elements that don't behave that way (e.g. a cubic Hermite element has 2^d dofs on each vertex, a Hierarchic element of degree p has p-1 dofs on each mid-edge node and p-1^2 on each mid-quad-face node) ... and I *thought* our infinite elements were among those!  Don't you end up with more and more dofs on a mid-infinite-edge node as you raise the polynomial degree of the radial basis functions use?

If you're trying to loop over nodes on a side, in general you want to forget dofs exist.  Either loop over all nodes and check `Elem::is_node_on_side` to decide which to skip, or get `Elem::nodes_on_side` and loop over those.

Reply:
Hi @roystgnr , I saw this only by now, sorry.
It is correct that infinite elements are not of "one dof index" = "one point" because the radial polynomials degree can be set independent from the element type. But I wouldn't think of those degrees as belonging to the 'mid-infinite points'. In fact, these mid points are not needed for the basis functions at all (given that the 'origin' is known; which in practice is computed using the outer points). 
Since both the order and the type of polynomial can be set to 'something', there are only very few properties one knows in general. But what they have in common is that the first one is a/r (where a is the distance of the 'inner' face from the origin) and the other polynomials start with 0 at the inner face. 
Thus, if I restrict the DOFs at nodes, I get those with `radial_dof=1` correct and restrict the other DOFs at the outer points, not knowing if the radial polynomials probably have a node there.

Therefore, I implemented the scheme slightly different: Knowing that the radial DOF of the basis functions with `radial_dof==1` is 1 at the base, I can restrict it in the same way as Lagrange elements (This function is only accessed when the non-radial (base) dofs are of Lagrange type).
DOFs with `radial_dof > 1` must be restricted in the same way as those with `radial_dof = 1` to ensure smoothness along the radial axis, i.e. if the first dof is restricted to 1/3 of its max. value, I must restrict all other DOFs that share the same base DOF to 1/3 of their actual maximum.

I hope, I made my point somehow understandable?
Probably this should be explained also in the sources in more detail...

Reply:
> I saw this only by now, sorry.

No worries.  Between "trying to wrap urgent things up by vacation time" and "vacation time" and "trying to catch up on what I missed during vacation time", I'm more behind on getting back to you on this ticket than vice-versa.

> But I wouldn't think of those degrees as belonging to the 'mid-infinite points'.

Where are their indices stored, then?  On the base points?

> Probably this should be explained also in the sources in more detail...

Yeah, even in the current version of the code I'm seeing `child_side->point(child_side_dof)`, and I still don't get it.  We do store all indices on some DofObjects, right, even in infinite elements?  And the number of indices can exceed the number of points, right?  So how can we look up a point by passing in a dof index rather than a point/node index?  How does that not become an OOB index when `n_elem_dofs` exceeds `n_points`?

Reply:
I hope the answer in the [the newest commit](https://github.com/libMesh/libmesh/pull/2792/commits/5a66cd08f6dbe8b7b6cbb863ccb3c80f56c426d8) helps to understand my idea.

>Where are their indices stored, then? On the base points?

Good point: No, they are indeed stored on the outer points. But this doesn't mean that these nodes are 'support' points: The radial polynomial may have a node or some small value there; so we may constrain them in a wrong way (or worse don't mark them for constraining when the radial polynomial has a node).

>  even in the current version of the code I'm seeing `child_side->point(child_side_dof)`

I think, the confusion starts with the name `child_side`; I have changed it to `child_base` now: independent of the `side` considered, we use the base face to compute the constraint; and there, we have `Lagrange`-type elements.
I hope this becomes clean from my new part in the documentation: Since the radial direction is not refined, here the basis functions have the same behavior and hence we are (in principle) free to choose a point where the constraints should be set.
But, instead of searching for an explicit point, I assume a point where the radial polynomial takes the value of 1, allowing me to compute the constraint just based on the `child_base`s functions.

Reply:
@roystgnr , I hope, with the newest explanation the function is easier to understand!?

Reply:
Sorry for taking so long to get back to this; I was banging my head against the early drafts of #2985, your variable name change and comment here didn't make things any more understandable at first glance, and before I found time to do another full read through I just let it drop off my attention.

At second glance ... it actually does make sense, doesn't it?  The infinite element DoFs and nodes are always ordered base-first, the base FE type is always Lagrange, and the Lagrange FE DoF indices are equal to local node indices, so of course you can interchange them?

I'd love it if we can eventually weaken "the base FE type is always Lagrange", but for now this looks great.

Reply:
>  The infinite element DoFs and nodes are always ordered base-first, the base FE type is always Lagrange

yes, this exactly is the point. 

I'd love it if we can eventually weaken "the base FE type is always Lagrange", but for now this looks great.

Yes, that indeed would be nice; but my impression was that most other elements don't compute constraints or use `compute_proj_constraints` which than require some generalizations; but I am not aware of how we could enter this part with non-Lagrange elements...

Reply:
`compute_proj_constraints` might not be as efficient as per-FE code (which, even though it probably wants to do just those projections, might be able to use more direct calculations rather than instantiating generic quadrature rules and inverting small systems), but it doesn't require much from the FE type and it gives the result we want for the space types it handles.

Can we run with base type != LAGRANGE on conforming meshes already?

Reply:
It seems that it works with Non-lagrange basis as well (I have tested Bernstein, xyz and monomial which seem to to be the only ones that work with Tet-elements) in the sense that I didn't receive errors.
I didn't check if the results make sense, though.